dspace-api/src/main/java/org/dspace/discovery/SolrServiceResourceRestrictionPlugin.java

```diff
@@ -61,8 +66,21 @@
 
     @Override
     public void additionalIndex(Context context, IndexableObject idxObj, SolrInputDocument document) {
+        DSpaceObject dso = null;
         if (idxObj instanceof IndexableDSpaceObject) {
-            DSpaceObject dso = ((IndexableDSpaceObject) idxObj).getIndexedObject();
+            dso = ((IndexableDSpaceObject) idxObj).getIndexedObject();
+        } else if (idxObj instanceof IndexableInProgressSubmission) {
+            final InProgressSubmission inProgressSubmission
+                    = ((IndexableInProgressSubmission) idxObj).getIndexedObject();
+            dso = inProgressSubmission.getItem();
+        } else if (idxObj instanceof IndexablePoolTask) {
+            final PoolTask poolTask = ((IndexablePoolTask) idxObj).getIndexedObject();
+            dso = poolTask.getWorkflowItem().getItem();
+        } else if (idxObj instanceof IndexableClaimedTask) {
+            final ClaimedTask claimedTask = ((IndexableClaimedTask) idxObj).getIndexedObject();
+            dso = claimedTask.getWorkflowItem().getItem();
```

Comment:
Not necessarily required of this PR, but this big if/else if seems like something we should refactor at some point.  It implies to me that `IndexableInProgressSubmission`, `IndexablePoolTask` and `IndexableClaimedTask` should all implement an interface that has a `getDSO()` (or `getRelatedDSO()`) method.   That would make this easier to extend in the future.

Reply:
I agree
We didn't want to change the discovery indexable objects in this PR, but it would definitely improve the functionality here, and can perhaps be used in the future in other locations where the item of any workflow object should be retrieved